content/doc/developer/FIPS-140/index.adoc

```diff
@@ -0,0 +1,47 @@
+---
+layout: developerchapter
+summary: FIPS-140 for developers
+wip: true
+title: FIPS-140 for developers
+references:
+- url: https://github.com/jenkinsci/jep/blob/master/jep/237/README.adoc
+  title: JEP-237 FIPS-140 support for Jenkins
+- url: hhttps://csrc.nist.gov/pubs/fips/140-2/upd2/final
+  title: FIPS-140-2 Security Requirements for Cryptographic Modules
+---
+
+In order for the code in Jenkins and plugins to be compliant with FIPS-140, sensitive information needs to be handled correctly.
+This document provides a non-exhaustive cheat sheet that can guide developers to write code that is FIPS-140 compliant.
+This applies to all forms of cryptography, regardless if it is encryption/decryption, hashing, or signing.
+[IMPORTANT]
+====
+FIPS-140 compliant code may result in less security as the approved standards can lag behind more secure modern standards.
+Your plugin and Jenkins code should not automatically follow the tips here, but may also need to make decisions at runtime whether or not to use a FIPS-140 compliant option.
+====
+
+Any sensitive information for accessing an external system that is persisted should be stored either as a link:https://javadoc.jenkins.io/hudson/util/Secret.html[`hudson.util.Secret`] or a link:https://javadoc.jenkins.io/plugin/credentials/com/cloudbees/plugins/credentials/Credentials.html[`com.cloudbees.plugins.credentials.Credentials`]
+
+Rely only on link:https://docs.oracle.com/en/java/javase/11/security/java-cryptography-architecture-jca-reference-guide.html#GUID-3E0744CE-6AC7-4A6D-A1F6-6C01199E6920[Java Cryptography Architecture (JCA)] or link:https://bouncycastle.org/java.html[BouncyCastle] via the link:https://plugins.jenkins.io/bouncycastle-api/[BouncyCastle API plugin] when performing cryptography.
+Do not depend on any third-party libraries or write your own code to perform encryption or specify any specific link:https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/security/Provider.html[`java.security.Provider`] when obtaining a concrete JCA implementation.
+Additionally, the subset of algorithms that are used should be restricted to the subset that is link:https://csrc.nist.gov/projects/cryptographic-module-validation-program/certificate/3514[supported] by the link:https://downloads.bouncycastle.org/fips-java/BC-FJA-UserGuide-1.0.2.pdf[BouncyCastle FIPS] implementation.
+[NOTE]
+====
+FIPS-140 has no support for asymmetric encryption such as RSA.
+====
+
+If there is a tradeoff made between FIPS-140 support or a more secure option, the tradeoff should be made at runtime by checking link:https://javadoc.jenkins.io/jenkins/security/FIPS140.html#useCompliantAlgorithms()[`jenkins.security.FIPS140.useCompliantAlgorithms`].
```

Comment:
This link won't resolve until Jenkins 2.426 is released and its Javadoc is available from https://javadoc.jenkins.io .

Reply:
which IIUC is tomorrow, so merging a day early for developer docs wouldn't seem to be the end of the world?

Reply:
Wholehearted agreement that it is perfectly fine to merge with this link temporarily broken.